instrumentation/jmx-metrics/library/src/test/java/io/opentelemetry/instrumentation/jmx/rules/HadoopTest.java

```diff
@@ -0,0 +1,177 @@
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.Test;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.strategy.Wait;
+import org.testcontainers.images.builder.Transferable;
+
+class HadoopTest extends TargetSystemTest {
+
+  public static final String ENDPOINT_PLACEHOLDER = "<<ENDPOINT_PLACEHOLDER>>";
+
+  @Test
+  void testMetrics_Hadoop2x() throws URISyntaxException, IOException {
+    List<String> yamlFiles = Collections.singletonList("hadoop.yaml");
+
+    yamlFiles.forEach(this::validateYamlSyntax);
+
+    // Hadoop startup script does not propagate env vars to launched hadoop daemons,
+    // so all the env vars needs to be embedded inside the hadoop-env.sh file
+    GenericContainer<?> target =
+        new GenericContainer<>("bmedora/hadoop:2.9-base")
+            .withCopyToContainer(
+                Transferable.of(readAndPreprocessEnvFile("hadoop2-env.sh")),
+                "/hadoop/etc/hadoop/hadoop-env.sh")
+            .withCreateContainerCmdModifier(cmd -> cmd.withHostName("test-host"))
+            .withStartupTimeout(Duration.ofMinutes(3))
+            .withExposedPorts(50070)
+            .waitingFor(Wait.forListeningPorts(50070));
+
+    copyAgentToTarget(target);
+    copyYamlFilesToTarget(target, yamlFiles);
+
+    startTarget(target);
+
+    verifyMetrics(createMetricsVerifier());
+  }
+
+  private String readAndPreprocessEnvFile(String fileName) throws URISyntaxException, IOException {
+    Path path = Paths.get(getClass().getClassLoader().getResource(fileName).toURI());
+
+    String data;
+    try (Stream<String> lines = Files.lines(path)) {
+      data =
+          lines
+              .map(line -> line.replace(ENDPOINT_PLACEHOLDER, otlpEndpoint))
+              .collect(Collectors.joining("\n"));
+    }
+
+    return data;
+  }
+
+  @Test
+  void testMetrics_Hadoop3x() throws URISyntaxException, IOException {
+    List<String> yamlFiles = Collections.singletonList("hadoop.yaml");
+
+    yamlFiles.forEach(this::validateYamlSyntax);
+
+    // Hadoop startup script does not propagate env vars to launched hadoop daemons,
+    // so all the env vars needs to be embedded inside the hadoop-env.sh file
+    GenericContainer<?> target =
+        new GenericContainer<>("loum/hadoop-pseudo:3.3.6")
+            .withExposedPorts(9870, 9000)
+            .withCopyToContainer(
+                Transferable.of(readAndPreprocessEnvFile("hadoop3-env.sh")),
+                "/opt/hadoop/etc/hadoop/hadoop-env.sh")
+            .withCreateContainerCmdModifier(cmd -> cmd.withHostName("test-host"))
+            .waitingFor(
+                Wait.forListeningPorts(9870, 9000).withStartupTimeout(Duration.ofMinutes(3)));
```

Comment:
[minor] this is the only part that differs between hadoop 2.x and 3.x, it could be worth refactoring this with a common method for the test body that delegates the container creation to a `Producer<GenericContainer<?>>` lambda with a dedicated implementation for each.

Reply:
Ok, then let's keep it that way until we need to deal with more variants.